cephclient: automate cephclient cephx key rotation

parth-gr opened the pull request:
<!-- Thank you for contributing to Rook! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Review our Contributing documentation at https://rook.io/docs/rook/latest/Contributing/development-flow/
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

BlaineEXE commented:
> do we also need to add `WithCephVersionUpgrade` in the enum?

Thanks for asking. We are not implementing this option with our first pass of feature additions.

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -3100,6 +3100,9 @@ type ClientSpec struct {
 	Caps map[string]string `json:"caps"`
+	// CephX configures CephX key settings. More: https://docs.ceph.com/en/latest/dev/cephx/
+	// +optional
+	CephX ClusterCephxConfig `json:"cephx,omitempty"`
```
The CephClient doesn't have a daemon associated, and the `ClusterCephxConfig` has a Daemon field. In the design doc, CephClient stands out in its simplicity, it should just have the raw cephx options.

```suggestion
	CephX CephxConfig `json:"cephx,omitempty"`
```

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -3100,6 +3100,17 @@ type ClientSpec struct {
 	Caps map[string]string `json:"caps"`
+	// Security represents security settings
+	// +optional
+	// +nullable
```
Not a pointer so not nullable

BlaineEXE reviewed (changes requested):
I don't see that the status is being updated. That is important, and the `reconcile()` method will need information about key rotation to properly set the status. 

You will likely need to follow the RGW example and pass `shouldRotate` into the `createOrUpdateClient()` method and determine `shouldRotate` beforehand.

Madhu-1 commented on pkg/operator/ceph/client/controller.go:
```diff
@@ -288,6 +304,38 @@ func (r *ReconcileCephClient) createOrUpdateClient(cephClient *cephv1.CephClient
 
+func (r *ReconcileCephClient) rotateCephXKeys(cephClient *cephv1.CephClient, user string) (string, error) {
+	// Detect desired CephCluster version
+	runningCephVersion, desiredCephVersion, err := currentAndDesiredCephVersion(
+		r.opManagerContext,
+		r.opConfig.Image,
+		cephClient.Namespace,
+		controllerName,
+		k8sutil.NewOwnerInfo(cephClient, r.scheme),
+		r.context,
+		r.clusterSpec,
+		r.clusterInfo,
+	)
```
would it make sense to run this job only when the policy is `WithCephVersionUpdate`?

BlaineEXE replied:
Thanks for bringing this up. We talked about this in huddle and agreed. We decided not to use the job for the current implementation.

parth-gr commented:
Testing: 

cr: 
```
spec:
  caps:
    mon: profile rbd
    osd: profile rbd pool=volumes, profile rbd pool=vms, profile rbd-read-only pool=images
  security:
    cephx: {}
status:
  cephx:
    keyCephVersion: 0.0.0-0
    keyGeneration: 1
  info:
    secretName: rook-ceph-client-cinder
  observedGeneration: 2
  phase: Ready
```

Secret:
` ``
apiVersion: v1
data:
  adminID: Y2luZGVy
  adminKey: QVFCQU9HVm9SdmFvQVJBQVB4c29GUWtwUFV3a3pibHFuQzBBWWc9PQ==
  cinder: QVFCQU9HVm9SdmFvQVJBQVB4c29GUWtwUFV3a3pibHFuQzBBWWc9PQ==
  userID: Y2luZGVy
  userKey: QVFCQU9HVm9SdmFvQVJBQVB4c29GUWtwUFV3a3pibHFuQzBBWWc9PQ==
kind: Secret
metadata:
  creationTimestamp: "2025-07-02T13:46:40Z"
  name: rook-ceph-client-cinder
  namespace: openshift-storage
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephClient
    name: cinder
    uid: 2110f1d8-8b6f-49e3-ba3d-75af86244587
  resourceVersion: "292095"
  uid: 5f5ecd1c-4679-4c0b-b71c-2b03428e312d
type: kubernetes.io/rook

```

ceph user:
```
client.cinder
	key: AQBAOGVoRvaoARAAPxsoFQkpPUwkzblqnC0AYg==
	caps: [mon] profile rbd
	caps: [osd] profile rbd pool=volumes, profile rbd pool=vms, profile rbd-read-only pool=images

```


After applying keyrotation spec:
```
spec:
  caps:
    mon: profile rbd
    osd: profile rbd pool=volumes, profile rbd pool=vms, profile rbd-read-only pool=images
  security:
    cephx:
      keyGeneration: 2
      keyRotationPolicy: KeyGeneration
status:
  cephx:
    keyCephVersion: 0.0.0-0
    keyGeneration: 1
  info:
    secretName: rook-ceph-client-cinder
  observedGeneration: 2
  phase: Ready
```

Secret
```
apiVersion: v1
data:
  adminID: Y2luZGVy
  adminKey: QVFEdE9HVm93ZUhqRGhBQWFRaThKSWkzVGNBY2F2eEFPZHFCSmc9PQ==
  cinder: QVFEdE9HVm93ZUhqRGhBQWFRaThKSWkzVGNBY2F2eEFPZHFCSmc9PQ==
  userID: Y2luZGVy
  userKey: QVFEdE9HVm93ZUhqRGhBQWFRaThKSWkzVGNBY2F2eEFPZHFCSmc9PQ==
kind: Secret
metadata:
  creationTimestamp: "2025-07-02T13:46:40Z"
  name: rook-ceph-client-cinder
  namespace: openshift-storage
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephClient
    name: cinder
    uid: 2110f1d8-8b6f-49e3-ba3d-75af86244587
  resourceVersion: "293491"
  uid: 5f5ecd1c-4679-4c0b-b71c-2b03428e312d
type: kubernetes.io/rook


````

ceph user:
```
client.cinder
	key: AQDtOGVoweHjDhAAaQi8JIi3TcAcavxAOdqBJg==
	caps: [mon] profile rbd
	caps: [osd] profile rbd pool=volumes, profile rbd pool=vms, profile rbd-read-only pool=images


```

BlaineEXE commented:
Please open a new PR for the change to switch usage for running/desired ceph version. As expected, I did validate in testing that I see the error.

Please also ensure the commit message resolves typos and follows guidelines.

```
core: return running and desired version from the version job

currently the version job was reurning desired and running version,
but all the function were using it oppositly
So update the version job to return it as per the correct order
in which it is consumed
```

Currently, the title doesn't give an accurate overview of the change. The function already returns running and desired version. This change switches the return value order.

The commit message should be clear, which I find to be best done by being specific about the function affected. Otherwise "version job" would be interpreted to mean only the version job, but this function uses that and the `ceph versions` output, which is a different layer. Ideally, I should be able to understand the major functional change only by reading the commit body.

This would be more clear in the details while also being just as short.

> core: correct return order for CurrentAndDesiredCephVersion()
>
> Currently CurrentAndDesiredCephVersion() returns desired and running
> version, but all function calls use the opposite return order. Update 
> the function to return the order as used by callers.

BlaineEXE commented on pkg/operator/ceph/client/controller.go:
```diff
@@ -268,6 +303,16 @@ func (r *ReconcileCephClient) createOrUpdateClient(cephClient *cephv1.CephClient
 
+	if shouldRotateCephxKeys {
+		// rotate the CephX key if the user requested it
+		rotatedKey, err := r.rotateCephXKeys(cephClient, clientEntity)
```
I don't think a function is necessary or helpful for 2 lines of code and a log message.

Also, because `keyring.SecretStore` isn't being used here today, it woudld be more straightforward to use the `cephclient.AuthRotate()` method, which uses the same abstraction layer as `cephclient.AuthGetOrCreateKey()` and `cephclient.AuthUpdateCaps()` above. That will reduce the lines of code to 1, further making a helper function unnecessary.

BlaineEXE commented on pkg/operator/ceph/client/controller_test.go:
It would be best to implement all of the `controller_test.go` [tests from RGW](https://github.com/rook/rook/pull/15813/files#diff-10bcb1ddae3b1fddd221dda0765d8dc3987af28dae3a1944dbeaadc31da531fd), translated to CephClient here. That will give us high confidence in cephx rotation changes for brown and greenfield CephClients via unit tests.

BlaineEXE commented on pkg/operator/ceph/client/controller.go:
```diff
@@ -227,28 +235,55 @@ func (r *ReconcileCephClient) reconcile(request reconcile.Request) (reconcile.Re
 		return reconcile.Result{}, *cephClient, errors.Wrapf(err, "failed to validate client %q arguments", cephClient.Name)
 	}
 
+	// Detect desired CephCluster version
+	runningCephVersion, desiredCephVersion, err := currentAndDesiredCephVersion(
+		r.opManagerContext,
+		r.opConfig.Image,
+		cephClient.Namespace,
+		controllerName,
+		k8sutil.NewOwnerInfo(cephClient, r.scheme),
+		r.context,
+		r.clusterSpec,
+		r.clusterInfo,
+	)
+	if err != nil {
+		if strings.Contains(err.Error(), opcontroller.UninitializedCephConfigError) {
+			logger.Info(opcontroller.OperatorNotInitializedMessage)
+			return opcontroller.WaitForRequeueIfOperatorNotInitialized, *cephClient, nil
+		}
+		return reconcile.Result{}, *cephClient, errors.Wrap(err, "failed to detect running and desired ceph version")
+	}
```
As discussed in huddle, since we aren't implementing `WithCephVersionUpdate`, this isn't necessary yet. I think we can replace this with `cephclient.LeastUptodateDaemonVersion()`, using the output for both running and desired to `ShouldRotateCephxKeys()`.

Add a comment to note that this will need adjusted when/if `WithCephVersionUpdate` is implemented. This may also be a matter of if it is implemented. CephClient doesn't benefit from `WithCephVersionUpdate` as much as CephCluster, CephFilesystem, and CephObjectStore, so we may choose not to implement it for CephClient unless requested by users.

BlaineEXE reviewed (changes requested):
The overall shape of the changes here looks good. I'd like to put more emphasis on unit tests to ensure correctness and prevent regressions. And as discussed in huddle, we can remove the version detect job invocation in this PR.

I mentioned in a comment, but we should have the `CurrentAndDesiredCephVersion()` return order fix as a separate PR. That is something we can consider backporting to 1.17 (and 1.16?) separately since it is a definite bug that could be affecting things beyond just key rotation.

parth-gr commented on pkg/operator/ceph/client/controller_test.go:
```diff
@@ -535,3 +552,257 @@ func TestReconcileCephClient_reconcileCephClientSecret(t *testing.T) {
+		assert.NotContains(t, secret.StringData["userKey"], "CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC==")
+	})
+
+	t.Run("rotate key - cephx status updated", func(t *testing.T) {
```
@BlaineEXE either this test work, or one above it work, 
Adding both together is keeping the reconcile not return.

BlaineEXE commented on pkg/operator/ceph/client/controller_test.go:
```diff
@@ -535,3 +552,257 @@ func TestReconcileCephClient_reconcileCephClientSecret(t *testing.T) {
+		scheme:           s,
+		context:          c,
+		opManagerContext: ctx,
+		recorder:         record.NewFakeRecorder(5),
```
```suggestion
		recorder:         record.NewFakeRecorder(10),
```

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -3100,6 +3100,16 @@ type ClientSpec struct {
+
+// ClinetSecuritySpec represents security settings for a Ceph Client
+type ClientSecuritySpec struct {
+	// CephX configures CephX key settings. More: https://docs.ceph.com/en/latest/dev/cephx/
```
I think this is relevant because many users won't have any idea what CephX is. This line is copy-pasted from what I set up for RGW, and for consistency I think it makes sense to keep

BlaineEXE commented on pkg/operator/ceph/client/controller.go:
```diff
@@ -375,7 +412,7 @@ func generateClientName(name string) string {
 
 // updateStatus updates an object with a given status
-func (r *ReconcileCephClient) updateStatus(observedGeneration int64, name types.NamespacedName, status cephv1.ConditionType) {
+func (r *ReconcileCephClient) updateStatus(observedGeneration int64, name types.NamespacedName, status cephv1.ConditionType, cephx *cephv1.CephxStatus) {
```
Let's wrap the internals of this in a `RetryOnConflict()` loop to help prevent possible double rotations if there is a status update conflict

BlaineEXE replied:
This is a corner case, but it won't be a common one. We have no better way of tracking rotation status at this time, so we have to live with the rare possiblity of double rotation.

BlaineEXE reviewed (changes requested):
Only one last thing that I missed in previous reviews. Looks good after allowing status update to retry

parth-gr commented:
I see two reconcile @travisn @BlaineEXE which should be fixed as a separate problem,

```
2025-07-10 13:55:33.189302 I | ceph-spec: resource "CephClient": "openshift-storage/cinder" spec has changed. diff=  v1.ClientSpec{
  	... // 2 identical fields
  	RemoveSecret: false,
  	Caps:         {"mon": "profile rbd", "osd": "profile rbd pool=volumes, profile rbd pool=vms, profile rbd-read"...},
  	Security: v1.ClientSecuritySpec{
  		CephX: v1.CephxConfig{
  			KeyRotationPolicy: "KeyGeneration",
- 			KeyGeneration:     5,
+ 			KeyGeneration:     10,
  		},
  	},
  }
2025-07-10 13:55:33.202924 I | ceph-spec: parsing mon endpoints: b=172.30.131.0:3300,c=172.30.59.194:3300,a=172.30.231.220:3300
2025-07-10 13:55:33.728291 I | ceph-client-controller: creating client cinder in namespace openshift-storage
2025-07-10 13:55:33.728320 I | cephclient: getting ceph auth key "client.cinder"
2025-07-10 13:55:34.243770 I | cephclient: updating ceph auth caps "client.cinder" to [osd profile rbd pool=volumes, profile rbd pool=vms, profile rbd-read-only pool=images mon profile rbd]
2025-07-10 13:55:34.797377 I | ceph-client-controller: rotating cephx key for CephClient openshift-storage/cinder
2025-07-10 13:55:34.797407 I | cephclient: rotating ceph auth key "client.cinder"
2025-07-10 13:55:35.350224 I | op-k8sutil: updated secret "openshift-storage/rook-ceph-client-cinder"
2025-07-10 13:55:35.350869 I | ceph-spec: controller will reconcile resource "rook-ceph-client-cinder" based on patch: [redacted patch details due to potentially sensitive content]
2025-07-10 13:55:35.373101 I | ceph-spec: parsing mon endpoints: b=172.30.131.0:3300,c=172.30.59.194:3300,a=172.30.231.220:3300
2025-07-10 13:55:35.895289 I | ceph-client-controller: creating client cinder in namespace openshift-storage
2025-07-10 13:55:35.895317 I | cephclient: getting ceph auth key "client.cinder"
2025-07-10 13:55:36.403602 I | cephclient: updating ceph auth caps "client.cinder" to [mon profile rbd osd profile rbd pool=volumes, profile rbd pool=vms, profile rbd-read-only pool=images]
2025-07-10 13:55:36.975522 I | op-k8sutil: updated secret "openshift-storage/rook-ceph-client-cinder"

```


We patch the cephclient, which goes an update the secret(1st reconcile), we have a predicate if we update a non-crd object it will reconcile, so as secret is updated we reconcile again(2nd reconcile)

BlaineEXE commented:
> We patch the cephclient, which goes an update the secret(1st reconcile), we have a predicate if we update a non-crd object it will reconcile, so as secret is updated we reconcile again(2nd reconcile)

I resolved this for RGW by creating an annotation that is ignored by predicates. This should be resolved by adding this annotation:

https://github.com/rook/rook/blob/5f50e85a251a68b4066eeb7442aaea1be2cd3bb9/pkg/operator/ceph/config/keyring/store.go#L39

should look like this

```
kind: Secret
metadata:
  annotations:
    cephx-keyring: ""
```

parth-gr commented:
Done cc @BlaineEXE

parth-gr commented on pkg/operator/ceph/client/controller.go:
```diff
@@ -227,28 +234,48 @@ func (r *ReconcileCephClient) reconcile(request reconcile.Request) (reconcile.Re
+		return reconcile.Result{}, *cephClient, errors.Wrapf(err, "failed to retrieve current ceph %q version", config.MonType)
+	}
+
+	// TODO: Revisit the parameters for ShouldRotateCephxKeys, when we implement WithCephVersionUpdate
```
Okay

BlaineEXE commented on pkg/operator/ceph/client/controller.go:
```diff
@@ -375,31 +414,41 @@ func generateClientName(name string) string {
+		return nil
+	})
+	if err != nil {
+		logger.Error(err)
```
Santosh discovered a bug earlier in RGW's update implementation. This should return an error.

parth-gr replied:
@sp98  would have fixed dont procceed if the key is not rotated https://github.com/rook/rook/pull/16091

In this check we are retrying to update the status, currently in the rook base we never return err on failed update of status,
1) If needed I can include the fail err for cephclient with this PR

2) Open a issue, where we can discuss on how to handle status update for all the controllers

BlaineEXE replied:
Please fix the issue here. Because status updates are critical for this feature, status update checks are a ripple effect requirement introduced by this work.

parth-gr replied:
Updated

BlaineEXE reviewed (changes requested):
Small fix for a bug scenario Santosh pointed out that also affects RGW

BlaineEXE commented:
Looks great! Thanks for the hard work
